File: game/combat/Bullet.lua
-- [ts]: Bullet.ts
local ____lualib = require("lualib_bundle") -- 1
local __TS__Class = ____lualib.__TS__Class -- 1
local __TS__ArraySetLength = ____lualib.__TS__ArraySetLength -- 1
local __TS__ArrayIndexOf = ____lualib.__TS__ArrayIndexOf -- 1
local __TS__New = ____lualib.__TS__New -- 1
local __TS__SetDescriptor = ____lualib.__TS__SetDescriptor -- 1
local ____exports = {} -- 1
local ____Dora = require("Dora") -- 2
local Color = ____Dora.Color -- 2
local DrawNode = ____Dora.DrawNode -- 2
local Node = ____Dora.Node -- 2
local Vec2 = ____Dora.Vec2 -- 2
local ____GameContext = require("game.core.GameContext") -- 4
local ctx = ____GameContext.ctx -- 4
local ____DamageSystem = require("game.combat.DamageSystem") -- 5
local DamageSystem = ____DamageSystem.DamageSystem -- 5
local ____ObjectPool = require("game.utils.ObjectPool") -- 6
local ObjectPool = ____ObjectPool.ObjectPool -- 6
local ____MathUtils = require("game.utils.MathUtils") -- 7
local distSq = ____MathUtils.distSq -- 7
local normalize = ____MathUtils.normalize -- 7
local scale = ____MathUtils.scale -- 7
local withAlpha = ____MathUtils.withAlpha -- 7
local ____Config = require("game.config.Config") -- 8
local Config = ____Config.Config -- 8
function ____exports.asEnemy(e) -- 11
	if e == nil then -- 11
		return nil -- 12
	end -- 12
	local enemy = e -- 13
	if enemy ~= nil and type(enemy.takeDamage) == "function" and enemy.isAlive then -- 13
		return enemy -- 15
	end -- 15
	return nil -- 17
end -- 11
____exports.Bullet = __TS__Class() -- 20
local Bullet = ____exports.Bullet -- 20
Bullet.name = "Bullet" -- 20
function Bullet.prototype.____constructor(self) -- 44
	self.active = false -- 28
	self.pos = Vec2.zero -- 29
	self.vel = Vec2.zero -- 30
	self.damage = 0 -- 31
	self.pierce = 0 -- 32
	self.split = 0 -- 33
	self.ricochet = 0 -- 34
	self.radius = 5 -- 35
	self.life = 0 -- 36
	self.color = 16777215 -- 37
	self.onSplit = nil -- 39
	self.root = Node() -- 40
	self.hitIds = {} -- 41
	self.node = DrawNode() -- 42
end -- 44
function Bullet.spawn(self, root, pos, vel, damage, range, radius, pierce, split, ricochet, color, onSplit) -- 47
	local b = ____exports.Bullet.bulletPool:acquire() -- 60
	b.active = true -- 61
	b.pos = pos -- 62
	b.vel = vel -- 63
	b.damage = damage -- 64
	b.pierce = pierce -- 65
	b.split = split -- 66
	b.ricochet = ricochet -- 67
	b.radius = radius -- 68
	local speed = vel.length -- 69
	b.life = speed > 0.001 and range / speed or 1 -- 70
	b.color = color -- 71
	b.root = root -- 72
	b.onSplit = onSplit ~= nil and (function(nb) return onSplit(nb) end) or nil -- 73
	__TS__ArraySetLength(b.hitIds, 0) -- 73
	b.node.position = Vec2(pos.x, pos.y) -- 75
	b.node.visible = true -- 76
	if b.node.parent ~= root then -- 76
		b.node:addTo(root) -- 79
	end -- 79
	return b -- 81
end -- 47
function Bullet.prototype.reset(self) -- 85
	self.active = false -- 86
	__TS__ArraySetLength(self.hitIds, 0) -- 86
	self.ricochet = 0 -- 88
	self.onSplit = nil -- 89
	self.node:clear() -- 90
	self.node.visible = false -- 92
end -- 85
function Bullet.prototype.recycle(self) -- 96
	if not self.active then -- 96
		return -- 97
	end -- 97
	____exports.Bullet.bulletPool:release(self) -- 98
end -- 96
function Bullet.prototype.update(self, dt, player) -- 102
	if not self.active then -- 102
		return -- 103
	end -- 103
	if player.homing > 0 then -- 103
		local ____this_1 -- 103
		____this_1 = ctx -- 105
		local ____opt_0 = ____this_1.findEnemiesNear -- 105
		local targets = ____opt_0 and ____opt_0(____this_1, self.pos, 180, 1) or ({}) -- 105
		local ____temp_2 -- 106
		if #targets > 0 then -- 106
			____temp_2 = ____exports.asEnemy(targets[1]) -- 106
		else -- 106
			____temp_2 = nil -- 106
		end -- 106
		local target = ____temp_2 -- 106
		if target ~= nil then -- 106
			local speed = self.vel.length -- 108
			local desired = normalize(Vec2(target.pos.x - self.pos.x, target.pos.y - self.pos.y)) -- 109
			local turn = math.min(1, dt * 7) -- 110
			local blended = normalize(Vec2(self.vel.x / speed * (1 - turn) + desired.x * turn, self.vel.y / speed * (1 - turn) + desired.y * turn)) -- 111
			self.vel = scale(blended, speed) -- 115
		end -- 115
	end -- 115
	self.pos = Vec2(self.pos.x + self.vel.x * dt, self.pos.y + self.vel.y * dt) -- 118
	self.life = self.life - dt -- 119
	local halfW = 1000 + Config.cullMargin -- 121
	local halfH = 1000 + Config.cullMargin -- 122
	if self.life <= 0 or math.abs(self.pos.x) > halfW or math.abs(self.pos.y) > halfH then -- 122
		self:recycle() -- 124
		return -- 125
	end -- 125
	local candidates = ctx.findEnemiesNear ~= nil and ctx:findEnemiesNear(self.pos, self.radius + 40, 8) or ({}) -- 128
	do -- 128
		local i = 0 -- 131
		while i < #candidates do -- 131
			do -- 131
				local enemy = ____exports.asEnemy(candidates[i + 1]) -- 132
				if enemy == nil or not enemy.isAlive or enemy.markedDead then -- 132
					goto __continue18 -- 133
				end -- 133
				if __TS__ArrayIndexOf(self.hitIds, enemy.id) >= 0 then -- 133
					goto __continue18 -- 134
				end -- 134
				local rr = self.radius + enemy.radius -- 135
				if distSq(self.pos, enemy.pos) <= rr * rr then -- 135
					local ____self_hitIds_3 = self.hitIds -- 135
					____self_hitIds_3[#____self_hitIds_3 + 1] = enemy.id -- 137
					self:onHit(enemy, player) -- 138
					if not self.active then -- 138
						return -- 139
					end -- 139
				end -- 139
			end -- 139
			::__continue18:: -- 139
			i = i + 1 -- 131
		end -- 131
	end -- 131
	self.node:clear() -- 143
	self.node.position = Vec2(self.pos.x, self.pos.y) -- 144
	self.node:drawDot( -- 145
		Vec2.zero, -- 145
		self.radius, -- 145
		Color(4278190080 | self.color) -- 145
	) -- 145
	local speed = math.sqrt(self.vel.x * self.vel.x + self.vel.y * self.vel.y) -- 146
	if speed > 0.001 then -- 146
		local nx = self.vel.x / speed -- 148
		local ny = self.vel.y / speed -- 149
		self.node:drawSegment( -- 150
			Vec2(-nx * self.radius * 0.5, -ny * self.radius * 0.5), -- 150
			Vec2(-nx * (self.radius + 16), -ny * (self.radius + 16)), -- 150
			math.max(2, self.radius * 0.7), -- 150
			Color(withAlpha(self.color, 105)) -- 150
		) -- 150
	end -- 150
	self.node:drawDot( -- 152
		Vec2.zero, -- 152
		math.max(2, self.radius * 0.42), -- 152
		Color(255, 248, 220, 245) -- 152
	) -- 152
end -- 102
function Bullet.prototype.onHit(self, enemy, player) -- 156
	local dir = normalize(self.vel) -- 157
	local info = DamageSystem:buildInfo( -- 158
		self.damage, -- 158
		player, -- 158
		"bullet", -- 158
		dir, -- 158
		"physical" -- 158
	) -- 158
	DamageSystem:apply(enemy, info) -- 159
	local ____opt_4 = ctx.vfx -- 159
	if ____opt_4 ~= nil then -- 159
		____opt_4:burst(enemy.pos, self.color, 6, 140) -- 160
	end -- 160
	local ____this_7 -- 160
	____this_7 = ctx -- 161
	local ____opt_6 = ____this_7.damageBreakablesInRadius -- 161
	if ____opt_6 ~= nil then -- 161
		____opt_6(____this_7, self.pos, self.radius + 12, info.amount) -- 161
	end -- 161
	if player.explosion > 0 then -- 161
		local ____this_9 -- 161
		____this_9 = ctx -- 163
		local ____opt_8 = ____this_9.damageEnemiesInRadius -- 163
		if ____opt_8 ~= nil then -- 163
			____opt_8(____this_9, enemy.pos, 52 + player.explosion * 8, { -- 163
				amount = self.damage * (0.35 + player.explosion * 0.1), -- 164
				kind = "magic", -- 165
				crit = false, -- 165
				knockback = Vec2.zero, -- 165
				hitStop = 0, -- 166
				shake = 2, -- 166
				flash = true, -- 166
				source = "explosion" -- 166
			}) -- 166
		end -- 166
		local ____opt_10 = ctx.vfx -- 166
		if ____opt_10 ~= nil then -- 166
			____opt_10:ring(enemy.pos, self.color, 52 + player.explosion * 8) -- 168
		end -- 168
	end -- 168
	if player.chain > 0 then -- 168
		local ____this_13 -- 168
		____this_13 = ctx -- 171
		local ____opt_12 = ____this_13.findEnemiesNear -- 171
		local chained = ____opt_12 and ____opt_12(____this_13, enemy.pos, 120, player.chain + 1) or ({}) -- 171
		do -- 171
			local i = 0 -- 172
			while i < #chained do -- 172
				local next = ____exports.asEnemy(chained[i + 1]) -- 173
				if next ~= nil and next.id ~= enemy.id then -- 173
					DamageSystem:apply(next, { -- 175
						amount = self.damage * 0.35, -- 176
						kind = "magic", -- 176
						crit = false, -- 176
						knockback = Vec2.zero, -- 177
						hitStop = 0, -- 177
						shake = 0, -- 177
						flash = true, -- 177
						source = "skill" -- 177
					}) -- 177
				end -- 177
				i = i + 1 -- 172
			end -- 172
		end -- 172
	end -- 172
	if self.ricochet > 0 then -- 172
		local ____this_15 -- 172
		____this_15 = ctx -- 183
		local ____opt_14 = ____this_15.findEnemiesNear -- 183
		local targets = ____opt_14 and ____opt_14(____this_15, enemy.pos, 180, 4) or ({}) -- 183
		do -- 183
			local i = 0 -- 184
			while i < #targets do -- 184
				local target = ____exports.asEnemy(targets[i + 1]) -- 185
				if target ~= nil and target.id ~= enemy.id and __TS__ArrayIndexOf(self.hitIds, target.id) < 0 then -- 185
					local speed = self.vel.length -- 187
					self.vel = scale( -- 188
						normalize(Vec2(target.pos.x - self.pos.x, target.pos.y - self.pos.y)), -- 188
						speed -- 188
					) -- 188
					self.pierce = math.max(self.pierce, 0) -- 189
					self.ricochet = self.ricochet - 1 -- 190
					return -- 191
				end -- 191
				i = i + 1 -- 184
			end -- 184
		end -- 184
	end -- 184
	if self.split > 0 and self.active then -- 184
		local baseAngle = math.atan(self.vel.y, self.vel.x) -- 197
		local speed = self.vel.length -- 198
		do -- 198
			local i = 0 -- 199
			while i < 2 do -- 199
				local a = baseAngle + (i == 0 and 0.45 or -0.45) -- 200
				local sub = ____exports.Bullet:spawn( -- 201
					self.root, -- 202
					self.pos, -- 203
					Vec2( -- 204
						math.cos(a) * speed, -- 204
						math.sin(a) * speed -- 204
					), -- 204
					self.damage, -- 205
					self.life * speed, -- 206
					self.radius * 0.8, -- 207
					self.pierce, -- 208
					self.split - 1, -- 209
					self.ricochet, -- 210
					self.color, -- 211
					self.onSplit -- 212
				) -- 212
				if self.onSplit ~= nil then -- 212
					self.onSplit(sub) -- 215
				end -- 215
				i = i + 1 -- 199
			end -- 199
		end -- 199
	end -- 199
	self.pierce = self.pierce - 1 -- 219
	if self.pierce < 0 then -- 219
		self:recycle() -- 221
	end -- 221
end -- 156
Bullet.bulletPool = __TS__New( -- 156
	ObjectPool, -- 22
	function() return __TS__New(____exports.Bullet) end, -- 23
	function(b) return b:reset() end -- 24
) -- 24
__TS__SetDescriptor( -- 24
	Bullet.prototype, -- 24
	"isActive", -- 24
	{get = function(self) -- 24
		return self.active -- 226
	end}, -- 226
	true -- 226
) -- 226
return ____exports -- 226

File: game/combat/Bullet.ts
// 弹道实体：移动、穿透、分裂，池化复用
import { Color, DrawNode, Node, Vec2 } from 'Dora';
import { DamageInfo, EnemyView } from 'game/core/Types';
import { ctx, PlayerView } from 'game/core/GameContext';
import { DamageSystem } from 'game/combat/DamageSystem';
import { ObjectPool } from 'game/utils/ObjectPool';
import { distSq, normalize, scale, withAlpha } from 'game/utils/MathUtils';
import { Config } from 'game/config/Config';

// 把 ctx.findEnemiesNear 返回的 unknown 窄化为 EnemyView
export function asEnemy(e: unknown): EnemyView | undefined {
	if (e === undefined) return undefined;
	const enemy = e as EnemyView;
	if (enemy !== undefined && typeof enemy.takeDamage === 'function' && enemy.isAlive) {
		return enemy;
	}
	return undefined;
}

export class Bullet {
	// 静态池
	static bulletPool = new ObjectPool<Bullet>(
		() => new Bullet(),
		(b) => b.reset()
	);

	// 状态
	active = false;
	pos: Vec2.Type = Vec2.zero;
	vel: Vec2.Type = Vec2.zero;
	damage = 0;
	pierce = 0;
	split = 0;
	ricochet = 0;
	radius = 5;
	life = 0;
	color = 0xffffff;
	// 分裂新子弹通知（由 WeaponSystem 注册，箭头包装）
	onSplit: ((b: Bullet) => void) | undefined = undefined;
	private root: Node.Type = Node();
	private hitIds: number[] = [];
	private node: DrawNode.Type = DrawNode();

	private constructor() {}

	// 从池中获取并初始化
	static spawn(
		root: Node.Type,
		pos: Vec2.Type,
		vel: Vec2.Type,
		damage: number,
		range: number,
		radius: number,
		pierce: number,
		split: number,
		ricochet: number,
		color: number,
		onSplit?: (b: Bullet) => void
	): Bullet {
		const b = Bullet.bulletPool.acquire();
		b.active = true;
		b.pos = pos;
		b.vel = vel;
		b.damage = damage;
		b.pierce = pierce;
		b.split = split;
		b.ricochet = ricochet;
		b.radius = radius;
		const speed = vel.length;
		b.life = speed > 0.001 ? range / speed : 1;
		b.color = color;
		b.root = root;
		b.onSplit = onSplit !== undefined ? (nb: Bullet) => onSplit(nb) : undefined;
		b.hitIds.length = 0;
		b.node.position = Vec2(pos.x, pos.y);
		b.node.visible = true;
		// 幂等挂载：池化节点复用时不重复 addTo
		if (b.node.parent !== root) {
			b.node.addTo(root);
		}
		return b;
	}

	// 归还池（重置）
	private reset(): void {
		this.active = false;
		this.hitIds.length = 0;
		this.ricochet = 0;
		this.onSplit = undefined;
		this.node.clear();
		// 池化节点不 remove（Dora remove 会 dispose 节点），隐藏等待复用
		this.node.visible = false;
	}

	// 回收（WeaponSystem 调用后回池）
	recycle(): void {
		if (!this.active) return;
		Bullet.bulletPool.release(this);
	}

	// 每帧更新：移动 + 碰撞
	update(dt: number, player: PlayerView): void {
		if (!this.active) return;
		if (player.homing > 0) {
			const targets = ctx.findEnemiesNear?.(this.pos, 180, 1) ?? [];
			const target = targets.length > 0 ? asEnemy(targets[0]) : undefined;
			if (target !== undefined) {
				const speed = this.vel.length;
				const desired = normalize(Vec2(target.pos.x - this.pos.x, target.pos.y - this.pos.y));
				const turn = Math.min(1, dt * 7);
				const blended = normalize(Vec2(
					this.vel.x / speed * (1 - turn) + desired.x * turn,
					this.vel.y / speed * (1 - turn) + desired.y * turn,
				));
				this.vel = scale(blended, speed);
			}
		}
		this.pos = Vec2(this.pos.x + this.vel.x * dt, this.pos.y + this.vel.y * dt);
		this.life -= dt;
		// 世界边界回收，不能按固定屏幕中心判断，否则玩家走远后无法射击
		const halfW = 1000 + Config.cullMargin;
		const halfH = 1000 + Config.cullMargin;
		if (this.life <= 0 || Math.abs(this.pos.x) > halfW || Math.abs(this.pos.y) > halfH) {
			this.recycle();
			return;
		}
		// 碰撞检测
		const candidates = ctx.findEnemiesNear !== undefined
			? ctx.findEnemiesNear(this.pos, this.radius + 40, 8)
			: [];
		for (let i = 0; i < candidates.length; i++) {
			const enemy = asEnemy(candidates[i]);
			if (enemy === undefined || !enemy.isAlive || enemy.markedDead) continue;
			if (this.hitIds.indexOf(enemy.id) >= 0) continue;
			const rr = this.radius + enemy.radius;
			if (distSq(this.pos, enemy.pos) <= rr * rr) {
				this.hitIds.push(enemy.id);
				this.onHit(enemy, player);
				if (!this.active) return;
			}
		}
		// 重绘
		this.node.clear();
		this.node.position = Vec2(this.pos.x, this.pos.y);
		this.node.drawDot(Vec2.zero, this.radius, Color(0xff000000 | this.color));
		const speed = Math.sqrt(this.vel.x * this.vel.x + this.vel.y * this.vel.y);
		if (speed > 0.001) {
			const nx = this.vel.x / speed;
			const ny = this.vel.y / speed;
			this.node.drawSegment(Vec2(-nx * this.radius * 0.5, -ny * this.radius * 0.5), Vec2(-nx * (this.radius + 16), -ny * (this.radius + 16)), Math.max(2, this.radius * 0.7), Color(withAlpha(this.color, 105)));
		}
		this.node.drawDot(Vec2.zero, Math.max(2, this.radius * 0.42), Color(255, 248, 220, 245));
	}

	// 命中敌人：结算伤害、穿透/分裂判定
	private onHit(enemy: EnemyView, player: PlayerView): void {
		const dir = normalize(this.vel);
		const info: DamageInfo = DamageSystem.buildInfo(this.damage, player, 'bullet', dir, 'physical');
		DamageSystem.apply(enemy, info);
		ctx.vfx?.burst(enemy.pos, this.color, 6, 140);
		ctx.damageBreakablesInRadius?.(this.pos, this.radius + 12, info.amount);
		if (player.explosion > 0) {
			ctx.damageEnemiesInRadius?.(enemy.pos, 52 + player.explosion * 8, {
				amount: this.damage * (0.35 + player.explosion * 0.1),
				kind: 'magic', crit: false, knockback: Vec2.zero,
				hitStop: 0, shake: 2, flash: true, source: 'explosion',
			});
			ctx.vfx?.ring(enemy.pos, this.color, 52 + player.explosion * 8);
		}
		if (player.chain > 0) {
			const chained = ctx.findEnemiesNear?.(enemy.pos, 120, player.chain + 1) ?? [];
			for (let i = 0; i < chained.length; i++) {
				const next = asEnemy(chained[i]);
				if (next !== undefined && next.id !== enemy.id) {
					DamageSystem.apply(next, {
						amount: this.damage * 0.35, kind: 'magic', crit: false,
						knockback: Vec2.zero, hitStop: 0, shake: 0, flash: true, source: 'skill',
					});
				}
			}
		}
		if (this.ricochet > 0) {
			const targets = ctx.findEnemiesNear?.(enemy.pos, 180, 4) ?? [];
			for (let i = 0; i < targets.length; i++) {
				const target = asEnemy(targets[i]);
				if (target !== undefined && target.id !== enemy.id && this.hitIds.indexOf(target.id) < 0) {
					const speed = this.vel.length;
					this.vel = scale(normalize(Vec2(target.pos.x - this.pos.x, target.pos.y - this.pos.y)), speed);
					this.pierce = Math.max(this.pierce, 0);
					this.ricochet -= 1;
					return;
				}
			}
		}
		// 分裂：向两侧斜向发射
		if (this.split > 0 && this.active) {
			const baseAngle = Math.atan2(this.vel.y, this.vel.x);
			const speed = this.vel.length;
			for (let i = 0; i < 2; i++) {
				const a = baseAngle + (i === 0 ? 0.45 : -0.45);
				const sub = Bullet.spawn(
					this.root,
					this.pos,
					Vec2(Math.cos(a) * speed, Math.sin(a) * speed),
					this.damage,
					this.life * speed,
					this.radius * 0.8,
					this.pierce,
					this.split - 1,
					this.ricochet,
					this.color,
					this.onSplit
				);
				if (this.onSplit !== undefined) {
					this.onSplit(sub);
				}
			}
		}
		this.pierce -= 1;
		if (this.pierce < 0) {
			this.recycle();
		}
	}

	get isActive(): boolean {
		return this.active;
	}
}
